lib/ex_doc/formatter/html/autolink.ex

```diff
@@ -370,112 +378,308 @@ defmodule ExDoc.Formatter.HTML.Autolink do
-    Regex.source(
-      ~r{([ct]:)?((#{module_re})?(([a-z_]+[A-Za-z_\d]*[\\?\\!]?)|[\{\}=&\\|\\.<>~*^@\\+\\%\\!-\/]+)/\d+)}
-    )
+  Ignores modules which are already wrapped in markdown url syntax,
+  e.g. `[Module](url)`. If the module name doesn't touch the leading
+  or trailing `]`, e.g. `[my link Module is here](url)`, the Module
+  will get translated to the new href of the module.
+  """
+  def erlang_modules(string) when is_binary(string) do
+    link(string, :erlang, :module, [])
+  end
 
-  @custom_re ~r{\[(.*?)\]\(`(#{fun_re})`\)}
-  @normal_re ~r{(?<!\[)`\s*(#{fun_re})\s*`(?!\])}
+  @doc """
+  Helper function for autolinking erlang functions.
 
-  defp replace_custom_links(bin, project_funs, extension, lib_dirs) do
-    Regex.replace(@custom_re, bin, fn all, text, match ->
-      replacement(all, match, project_funs, extension, lib_dirs, text)
-    end)
+  Only links modules that are in the Erlang distribution `lib_dir`
+  and only link functions in those modules that export a function of the
+  same name and arity.
+
+  Ignores functions which are already wrapped in markdown url syntax,
+  e.g. `[:module.test/1](url)`. If the function doesn't touch the leading
+  or trailing `]`, e.g. `[my link :module.link/1 is here](url)`, the :module.fun/arity
+  will get translated to the new href of the function.
+  """
+  def erlang_functions(string) when is_binary(string) do
+    link(string, :erlang, :function, [])
+  end
+
+  defp replace_fun(language, kind, link_type, options) do
+    case link_type do
+      :normal ->
+        fn all, match ->
+          replacement(all, language, kind, match, options)
+        end
+
+      :custom ->
+        fn all, text, match ->
+          replacement(all, language, kind, match, text, options)
+        end
+    end
+  end
+
+  # Helper function for autolinking functions and modules.
+  #
+  # It autolinks all links for a certain `language` and of a certain `kind`.
+  #
+  # `language` can be: `:elixir`, `:erlang` or `:markdown`.
+  #
+  # `kind` is either `:function` or `:module`.
+  #
+  # It accepts a list of `options` used in the replacement functions.
+  # - `:aliases
+  # - `:docs_refs`
+  # - `:extension` - Default value is `".html"`
+  # - `:lib_dirs`
+  # - `:locals` - A list of local functions
+  # - `:module_id` - Module of the current doc. Default value is `nil`
+  # - `:modules_refs` - List of modules available
+  #
+  # Internal options:
+  # - `:preprocess?` - `true` or `false`. Do preprocessing and postprocessing, such as replacing backticks
+  #                     with a token
+  @doc false
+  @spec link(String.t(), language, kind, keyword) :: String.t()
+  def link(string, language, kind, options) do
+    options = Keyword.put_new(options, :preprocess?, true)
+
+    string
+    |> link_process(:preprocess, options[:preprocess?])
+    |> link(language, kind, :custom, options)
+    |> link(language, kind, :normal, options)
+    |> link_process(:postprocess, options[:preprocess?])
```

Comment:
Wrong option here. It seems we do not have tests for this then?

Reply:
Sorry, what is wrong here?